content/article/why-i-hate-json-and-you-can-too-.md

```diff
@@ -0,0 +1,177 @@
+
+  {
+    "title"       : "Why I Hate JSON (and you can, too)",
+    "authors"     : ["felipe-gasper"],
+    "date"        : "2019-11-10T21:52:51",
+    "tags"        : [],
+    "draft"       : true,
+    "image"       : "",
+    "thumbnail"   : "",
+    "description" : "Another look at this ubiquitous serialization …",
+    "categories"  : "data"
+  }
+
+[JSON](https://www.json.org/) needs no introduction among most programmers.
+Since its “discovery” by Douglas Crockford in the early 2000s it has become
+one of the most popular data serialization formats yet created, having
+effectively supplanted XML as the web’s “go-to” format for representing
+structured data in APIs.
+
+For all its popularity, though, JSON carries a number of liabilities that
+make it less than ideal in some contexts where it nonetheless enjoys
+widespread use. This article will discuss some of these and propose
+alternative formats that I think deserve your attention.
+
+
+# 1. UTF-8
+
+JSON’s chief liability is its UTF-8 restriction: all JSON documents must
+be UTF-8, from start to finish. This means any application that wants to
+exchange binary data either cannot use JSON, or—perhaps more commonly—uses
+a secondary encoding like Base64 to store binary strings as text.
+Such techniques carry performance penalties and eventually create consistency
+problems, such as in the
+[WAMP](https://wamp-proto.org/_static/gen/wamp_latest_ietf.html) protocol,
+where it’s impossible to distinguish an “encoded” binary
+string from a text string that happens to fit the pattern that the protocol
+recommends for encoding binary strings to JSON.
+
+UTF-8 is a fine way of encoding text, but applications that use JSON bind
+themselves to a restriction that may prove to have been imprudent—sometimes
+only long after an ecosystem has already developed and a serialization change
+would be painful.
+
+The most bizarre aspect of this restriction is that it is wholly
+superfluous: the only reason why JSON strings must conform to UTF-8 is because
+its definition says so. The encoding itself could notate any
```

Comment:
If you want to include this paragraph, give some space to the design decision that led to this. There's some interesting stuff in here: http://doc.cat-v.org/bell_labs/utf-8_history

Reply:
I don’t see anything about JSON in here … am I missing something? This seems to be a generic history of UTF-8 specifically.

What I mean in this paragraph is that this part of the JSON spec is fully self-limiting. The encoding itself is syntactically capable of handling non-UTF-8.

Reply:
UTF-8 was designed to be able to recognize a corruption in the encoding and move to the next known good spot. That's an important aspect that other encodings can't do.